src/lammpsio/snapshot.py

```diff
@@ -176,31 +222,122 @@ def to_hoomd_gsd(self, type_map=None):
                 for typeidx, typeid in enumerate(sorted_typeids):
                     frame.particles.typeid[self.typeid == typeid] = typeidx
             else:
-                frame.particles.types = list(type_map.values())
-                reverse_type_map = {
-                    typeid: typeidx for typeidx, typeid in enumerate(type_map.keys())
-                }
-                for i, typeid in enumerate(self.typeid):
-                    frame.particles.typeid[i] = reverse_type_map[typeid]
+                if type_map is not None:
+                    warnings.warn(
+                        "type_map will be deprecated, use label instead.",
+                        DeprecationWarning,
+                        stacklevel=2,
+                    )
+                    type_label_map = LabelMap(map=type_map)
+                else:
+                    type_label_map = self.label
+                frame.particles.types = type_label_map.types
+                Snapshot._set_type_id(
+                    self,
+                    self.typeid,
+                    frame.particles.typeid,
+                    self.label,
+                )
         if self.has_charge():
             frame.particles.charge = self.charge.copy()
         if self.has_mass():
             frame.particles.mass = self.mass.copy()
         if self.has_molecule():
             frame.particles.body = self.molecule - 1
+        if self.has_bonds():
+            frame.bonds.N = self.bonds.N
+            frame.bonds.types = self.bonds.id
```

Comment:
I'm confused by this. It doesn't seem like we should assign `bonds.N` types as the default behavior.

If bonds has typeids, then you have two cases:
* There is a label map, so the types come from the label map.
* There is no label map, so the types are the unique typeids.

If bonds does not have typeids, you should not write them. It will be an ill-formed GSD file, but that is not our problem.

Reply:
That makes sense. I think this should raise an error letting users know that they must specify typids to convert to GSD.

Reply:
I don’t think it has to be an error since it returns a Frame that the caller could decide to modify. GSD should error out itself if a bad Frame gets written to file

Reply:
Should it be a warning? I feel like it could be pretty annoying to try to debug why your topology isn't converting while everything seemingly works fine.

Reply:
What about forcing a call to `frame.validate()`? That way GSD will internally self-validate:

https://gsd.readthedocs.io/en/v3.3.2/python-module-gsd.hoomd.html#gsd.hoomd.Frame.validate

I think this method ignores things that are `None`, so that seems to be allowed behavior and I think we should respect that.